### DISCUSSION TITLE
I always forget to add .scalars() at the end of my ORM queries

### DISCUSSION THREAD

**Discussion by miguelgrinberg** [sqlalchemy/sqlalchemy#6985] [Ideas]
Hi!

I'm starting to use SQLAlchemy 1.4 with the new session/query interfaces, after having used the older queries for years. The vast majority of my queries are for a single ORM entity, so I find the need to stick a `.scalars()` at the end of almost every query a bit frustrating, because I always forget to add it.

Fooling around with the `Session` class I came up with the following customized class that automatically does a scalar conversion when the result has only one column:

```python
from sqlalchemy.orm import Session as BaseSession

class Session(BaseSession):
    def execute(self, *args, **kwargs):
        auto_scalars = kwargs.pop('auto_scalars', True)
        result = super().execute(*args, **kwargs)
        if auto_scalars and len(result.keys()) == 1:
            return result.scalars()
        return result
```

I was wondering why something like this isn't part of the `Session` class, given that the old `Query` object did this. Are there any issues with this pattern that made you decide it is best to require an explicit conversion to scalar?

Why do I ask? I'm writing a small wrapper for SQLAlchemy that is somewhat based on the ideas introduced by the Flask-SQLAlchemy extension for Flask. My wrapper, however, is not tied to Flask, and it uses the new 1.4/2.x session and query style. I'm wondering if a solution like the above has some hidden problems I'm not seeing, if I were to include it in my wrapper. Thanks!

**Comment by zzzeek** [comment]
how about session.scalars(), to go with session.scalar() ?

**Reply by zzzeek** [reply]
we try to stay off keyword args for major "return value" types of things.  in this case, the return type of session.execute() is Result, whereas for session.scalars() it will be [ScalarResult](https://docs.sqlalchemy.org/en/14/core/connections.html?highlight=scalarresult#sqlalchemy.engine.ScalarResult).  Im very glad I went through the effort to create these separate result classes as they work out well for typing.

**Reply by miguelgrinberg** [reply]
Thanks for your reply, Mike. I do see the benefit in having clean return types to help linters/IDEs/etc.

A `session.scalars()` would be awesome, much better than my hack! Is this something you'd be interested in adding to the project, or should I still think about fitting it into my wrapper in a subclassed session?

**Reply by CaselIT** [reply]
I'm fine either way, the only issue may be that people may confuse it with scalar, but I think it's not too big of an issue since the same method is also in the result class, so people should already be familiar with it.

Would this also go on the connection?

**Reply by miguelgrinberg** [reply]
Pulll request submitted #6991. I have added it to the engine and session classes both. Also added a `stream_scalars()` method on the asyncio classes for consistency. Let me know how this looks!

**Comment by CaselIT** [comment]
> I was wondering why something like this isn't part of the Session class, given that the old Query object did this

Iirc the idea was to remove magic from sqlalchemy working.
The idea is: execute returns a tuple like objects in all cases, not sometimes a tuples and sometimes not

**Reply by miguelgrinberg** [reply]
Yep, makes perfect sense now that you guys mention it. :)
